### PR TITLE
feat(forge): Check tag in forge install

### LINKED ISSUE #1248: Warn on falling back to master tag when installing dependencies with errant tag
### Component

Forge

### Describe the feature you would like

This item is perhaps a minor defect, or a small improvement that may save someone else some confusion. 

I had a contract with imports like:

```solidity
import "@openzeppelin/contracts@3.4.0/...";
```

I ran
```bash
forge install OpenZeppelin/openzeppelin-contracts@3.4.0 
```

This ran and (seemingly) worked. The contract wouldn't build due to version contracts.

After some confusion, I read the `/lib/openzeppelin-contracts/package.json` which showed the version as `4.5.0` which is the current **master** version.

```json
{
  "name": "openzeppelin-solidity",
  "description": "Secure Smart Contract library for Solidity",
  "version": "4.5.0",
   ...
```

I went to github and found the tag which was [**v**3.4.0.](https://github.com/OpenZeppelin/openzeppelin-contracts/releases/tag/v3.4.0) (note the "v" that was absent in the original install tag version)

I tried an intentionally errant tag (note the "_asdf_" at the end) and again got the master `4.5.0` version. 
 ```bash
forge install OpenZeppelin/openzeppelin-contracts@v3.4.0asdf 
```

After successfully removing the errant version the correct @tag installed successfully. The version matched in the package.json, and all contracts built:

 ```bash
forge install OpenZeppelin/openzeppelin-contracts@v3.4.0
```

IMO either fail to install or warn that the requested tag was not installed but a fallback version.

***
#### `forge remove` left gitmodule issues that required post-hoc cleanup and manual handling

After running `forge remove OpenZeppelin/openzeppelin-contracts`, there was an error when trying to `install` again (I didn't record this exact error). The error message essentially said there was already a module `OpenZeppelin/openzeppelin-contracts`.

I _basically_ followed this top answer suggested [here on SO](https://stackoverflow.com/questions/1260748/how-do-i-remove-a-submodule). I repeated this a few times while trying/removing, etc. 

The remove process could also clean up the git modules (maybe?!). That seems a bit kludgy I realise, but I'm not sure what else to recommend.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by pyk**
## Motivation

Fix #1248 

## Solution

It checks the reference on remote repository via `git ls-remote` before running the installation

```
✦ forgedev install Rari-Capital/solmate@v7
Installing solmate in "/Users/pyk/github/pyk/foundry-install/lib/solmate", (url: https://github.com/Rari-Capital/solmate, tag: Some("v7"))
Error:
   0: tag/branch/commit "v7" is not exists

Location:
   cli/src/cmd/forge/install.rs:114

Backtrace omitted. Run with RUST_BACKTRACE=1 environment variable to display it.
Run with RUST_BACKTRACE=full to include source snippets.
```

### CODE REVIEW

File: cli/src/cmd/forge/install.rs [hunk 1]
```diff
@@ -101,6 +102,21 @@ pub(crate) fn install(
     Ok(())
 }
 
+/// make sure tag is exists on the remote repository
```
Comment (mattsse) [inline cli/src/cmd/forge/install.rs:105]:
```suggestion
/// make sure tag exists on the remote repository
```

File: cli/src/cmd/forge/install.rs [hunk 2]
```diff
@@ -101,6 +102,21 @@ pub(crate) fn install(
...
+fn check_tag(dep: &Dependency) -> eyre::Result<()> {
+    if let Some(ref tag) = dep.tag {
+        let output = Command::new("git")
+            .args(&["ls-remote", &dep.url, tag])
+            .stdout(Stdio::piped())
+            .output()?;
+        let stdout = str::from_utf8(&output.stdout).unwrap();
```
Comment (mattsse) [inline cli/src/cmd/forge/install.rs:112]: let's do lossy here: `String::from_utf8_lossy`

File: cli/src/cmd/forge/install.rs [hunk 3]
```diff
@@ -101,6 +102,21 @@ pub(crate) fn install(
...
+        if !stdout.contains(tag) {
+            eyre::bail!("tag/branch/commit \"{}\" is not exists", tag)
```
Comment (mattsse) [inline cli/src/cmd/forge/install.rs:114]:
```suggestion
            eyre::bail!("tag/branch/commit \"{}\" does not exists", tag)
```

**Review by mattsse — CHANGES_REQUESTED** [review]
thanks!
smol nits :)

**Comment by pyk** [comment]
@mattsse updated, please check

**Review by mattsse — APPROVED** [review]

**Pull request merged**